implementations/rust/ockam/ockam_transport_tcp/src/workers/sender.rs

```diff
@@ -58,120 +47,164 @@ pub(crate) enum TcpSendWorkerMsg {
-    pub(crate) async fn new_pair(
-        router_handle: TcpRouterHandle,
-        stream: Option<TcpStream>,
+    pub(crate) async fn create(
+        registry: TcpRegistry,
+        stream: TcpStream,
+        role: ConnectionRole,
         peer: SocketAddr,
-        hostnames: Vec<String>,
-    ) -> Result<(Self, WorkerPair)> {
-        let role_str = if stream.is_none() {
-            "initiator"
-        } else {
-            "responder"
-        };
-
-        let tx_addr = Address::random_tagged(&format!("TcpSendWorker_tx_addr_{}", role_str));
-        let int_addr = Address::random_tagged(&format!("TcpSendWorker_int_addr_{}", role_str));
-        let rx_addr = Address::random_tagged(&format!("TcpRecvProcessor_{}", role_str));
-        let sender = TcpSendWorker::new(router_handle, stream, peer, int_addr, rx_addr);
-        Ok((
-            sender,
-            WorkerPair {
-                hostnames,
-                peer,
-                tx_addr,
-            },
-        ))
+        outgoing_access_control: Arc<dyn OutgoingAccessControl>,
+    ) -> Result<Self> {
+        let role_str = role.str();
+
+        let self_address = Address::random_tagged(&format!("TcpSendWorker_tx_addr_{}", role_str));
+        let self_internal_address =
+            Address::random_tagged(&format!("TcpSendWorker_int_addr_{}", role_str));
+        let receiver_processor_address =
+            Address::random_tagged(&format!("TcpRecvProcessor_{}", role_str));
+        let sender = TcpSendWorker::new(
+            registry,
+            stream,
+            peer,
+            self_internal_address,
+            self_address,
+            receiver_processor_address,
+            outgoing_access_control,
+        );
+        Ok(sender)
     }
 
-    /// Start a `(TcpSendWorker, TcpRecvProcessor)` pair that opens and
+    /// Create a `(TcpSendWorker, TcpRecvProcessor)` pair that opens and
     /// manages the connection with the given peer
-    pub(crate) async fn start_pair(
+    pub(crate) async fn start(
         ctx: &Context,
-        router_handle: TcpRouterHandle,
-        stream: Option<TcpStream>,
+        registry: TcpRegistry,
+        stream: TcpStream,
+        role: ConnectionRole,
         peer: SocketAddr,
-        hostnames: Vec<String>,
-    ) -> Result<WorkerPair> {
-        let tcprouter_main_addr = router_handle.main_addr().clone();
-
+        sender_incoming_access_control: Arc<dyn IncomingAccessControl>,
+        receiver_outgoing_access_control: Arc<dyn OutgoingAccessControl>,
+    ) -> Result<Address> {
         trace!("Creating new TCP worker pair");
-        let (sender, pair) = Self::new_pair(router_handle, stream, peer, hostnames).await?;
-
-        // Allow messages routed from Tcp Router
-        let tx_mailbox = Mailbox::new(
-            pair.tx_addr(),
-            Arc::new(AllowSourceAddress(tcprouter_main_addr)),
+        let sender_worker = Self::create(
```

Comment:
This is a bit confusing. We create a `sender_worker` and we pass it a `receiver_outgoing_access_control`. Is it a typo? No, because the `TcpSendWorker` actually creates and starts a `TcpRecvProcessor` inside its `initialize` method. 

Why don't we do both the creation of the sender and the receiver inside of the `start` function, and use the required access control there? There will be more symmetry in the code and less indirection IMO.

Reply:
Nice catch. That was impossible before I moved TcpStream creation out of TcpSendWorker. Now should be possible

Reply:
I'll do it in another PR